### DISCUSSION TITLE
Can Spring Boot Extensions be used with this plugin?

### DISCUSSION THREAD

**Discussion by mikehaertl** [mfussenegger/nvim-jdtls#433] [Q&A]
I'm still new to Java so bear with me but I wondered if there's a way to also use  [Spring Boot Extension](https://marketplace.visualstudio.com/items?itemName=vmware.vscode-boot-dev-pack) for vscode with this plugin. My goal is e.g. to get LSP completions also for `application.properties` files.

In a naive approach I tried the following:

* Download und unzip [vscode-spring-boot-1.44.0-RC.2.vsix](https://github.com/spring-projects/sts4/releases/download/4.17.2.RELEASE/vscode-spring-boot-1.44.0-RC.2.vsix)
* Add the path to the unzipped `extensions/jars/*.jar` to the `bundles` in jdtls config
* Copy the `ftplugin/java.lua` to `ftplugin/jproperties.lua`

This at least starts the LSP now for `application.properties`. But I get no completions or other LSP features.

Is there a way to make this work?

EDIT: Somewhat related discussion for emacs' lsp-java: https://github.com/spring-projects/sts4/issues/76

**Comment by EpsilonKu** [comment]
I guess u need to setup it like standard lsp not like jdtls. I think it's possible with null-ls.

**Comment by mfussenegger** [comment]
Adding the jar files to the `bundles` should work in that it loads the eclipse.jdt.ls extensions of the Spring Boot Extension.

You can verify that by running:

```
lua =vim.lsp.get_active_clients({name = 'jdtls'})[1].server_capabilities.executeCommandProvider
```

You should see some commands that start with `sts`.

But I'm not sure if the Spring Boot extension actually uses jdtls.ls for `application.properties` files. Could be that that's done via another language server or via a vscode specific functionality.

And using the functionality added by the extra jar files probably also requires some client side code as well.

**Reply by mikehaertl** [reply]
> You should see some commands that start with sts.

It required me to play around with the order of the jar files but this now seems to do the trick:

```
vim.list_extend(
  bundles,
  {
    nvim_data_path .. 'extension/jars/io.projectreactor.reactor-core.jar',
    nvim_data_path .. 'extension/jars/org.reactivestreams.reactive-streams.jar',
    nvim_data_path .. 'extension/jars/jdt-ls-commons.jar',
    nvim_data_path .. 'extension/jars/jdt-ls-extension.jar',
  }
)
```

Now I see some commands:

```
 "sts.java.addClasspathListener",
 "sts.java.code.completions",
 "sts.java.hierarchy.subtypes",
 "sts.java.hierarchy.supertypes",
 "sts.java.javadoc",
 "sts.java.javadocHoverLink",
 "sts.java.location",
 "sts.java.removeClasspathListener",
 "sts.java.search.packages",
 "sts.java.search.types",
 "sts.java.type",
```

But still no LSP features.

> But I'm not sure if the Spring Boot extension actually uses jdtls.ls for application.properties files. Could be that that's done via another language server or via a vscode specific functionality.
>
> And using the functionality added by the extra jar files probably also requires some client side code as well.

I see.  From [this comment](https://github.com/spring-projects/sts4/issues/76#issuecomment-427074764) it seems that the client has to register to `sts.java.addClasspathListener` and `sts.java.removeClasspathListener`. But TBH I'm not even sure what that means exactly.

Looking at the emacs implementation in their lsp-java client I think it's not that much code required. But it's above my head TBH: https://github.com/emacs-lsp/lsp-java/blob/master/lsp-java-boot.el

It would of course be awesome if this could be integrated into nvim-jdtls.
